File: 2015/03/part_2.rb
# Usage: From root folder
# $ ruby 2015/03/part_2.rb 2015/03/input

class Position
  def initialize
    @x, @y = [0, 0]
  end

  def get_key
    "#{@x}#{@y}"
  end

  def move direction
    case direction
    when "<"
      @x-=1
    when ">"
      @x+=1
    when "^"
      @y+=1
    when "v"
      @y-=1
    end
  end
end

File.open(ARGV[0]).readlines.each do |directions|
  santa  = Position.new
  robo_santa  = Position.new
  houses = {santa.get_key => true}

  directions.split("").each_slice(2) do |dir_santa, dir_robo|
    santa.move dir_santa
    houses[santa.get_key] = true

    robo_santa.move dir_robo
    houses[robo_santa.get_key] = true
  end

  puts(houses.size)
end

File: 2015/03/part_2.py
#  Usage: From root folder
#  $ python 2015/03/part_2.py 2015/03/input

import sys


class Position:
    def __init__(self):
        self.x, self.y = 0, 0

    def get_key(self):
        return f"{self.x}{self.y}"

    def move(self, direction):
        if direction == "<":
            self.x -= 1
        elif direction == ">":
            self.x += 1
        elif direction == "^":
            self.y += 1
        elif direction == "v":
            self.y -= 1


if __name__ == "__main__":
    with open(sys.argv[1]) as file:
        for directions in file.readlines():
            santa = Position()
            robo_santa = Position()
            houses = {santa.get_key(): True}

            for direction in directions.strip()[0::2]:
                santa.move(direction)
                houses[santa.get_key()] = True

            for direction in directions.strip()[1::2]:
                robo_santa.move(direction)
                houses[robo_santa.get_key()] = True

            print(len(houses))
